README.md

# App Update alert Plugin
```keep your app users up to date on their devices```

[![ios support version - 13+](https://img.shields.io/badge/ios_support_version-13%2B-2ea44f)](https://)
[![android support version - 8+](https://img.shields.io/badge/android_support_version-8%2B-2ea44f)](https://)
![tag - v1.5.0](https://img.shields.io/badge/tag-v1.5.0-516cc4?logo=3a76b8)
[![License](https://img.shields.io/badge/License-Apache_2.0-blue)](#license)
[![News - Android Weekly](https://img.shields.io/badge/News-Android_Weekly-d36f21)](https://androidweekly.net/issues/issue-326)
[![Story - Medium](https://img.shields.io/badge/Story-Medium-2ea44f)](https://medium.com/codex/image-compresso-13dbfd0445a3)
[![GitHub - VinodBaste](https://img.shields.io/badge/GitHub-VinodBaste-516cc4)](https://github.com/vinodbaste/app_update#readme)

With the help of a this Flutter plugin.
* Check if a user has the most recent installation of your app.
* Display a message to the user with a link to the relevant app store page.

See more at the [Dart Packages page.](https://pub.dev/packages/app_update_alert)

<img src = "https://github.com/vinodbaste/app_update/raw/main/screenshots/android.png" width = 250 height = 500 />

## Installation
Add `app_update_alert` as [a dependency in your `pubspec.yaml` file.](https://flutter.io/using-packages/)

## With Flutter:
```
Run this command:
flutter pub add app_update_alert
```
This will add a line like below to your package's ```pubspec.yaml```

```
dependencies:
  app_update_alert: ^1.5.5
```
run an implicit ```flutter pub get```

## Usage
Create an instance of `AppUpdate` in `main.dart` (or wherever your app is initialised).

```Dart
@override
  void initState() {
    super.initState();
 final appUpdate = AppUpdate(
      iosPackageName: 'com.your.IOSpackage',
      androidPackageName: 'com.your.Androidpackage', 
      iOSAppStoreCountry: 'in'
  );
}
```

Your Flutter package identification will be used by the plugin automatically to search the app store.
You can overwrite this identity if your app uses a different one on the Google Play Store or Apple App Store by giving values for `androidPackageName` and/or `iosPackageName.`

*For iOS:* You must specify `iOSAppStoreCountry` to the two-letter country code of the Software Store you want to search if your app is only accessible outside the India. A list of ISO Country Codes can be found at http://en.wikipedia.org/wiki/ISO 3166-1 alpha-2.


### Quickstart
Calling `showUpdateAlert` with your app's `BuildContext` and other optional parameters which will check if the app can be updated, and will automatically display a platform-specific alert that the user can use to go to the app store.

```Dart
@override
  void initState() {
    super.initState();
 final appUpdate = AppUpdate(
      iosPackageName: 'com.your.IOSpackage',
      androidPackageName: 'com.your.Androidpackage', 
      iOSAppStoreCountry: 'in'
  );
  
 //showUpdateAlert
 appUpdate.showUpdateAlert(context: context);
}
```
*Note:* The parameters such as `iosPackageName`,  `androidPackageName` are non-mandatory fields and can only be overwritten if the app package's differ from store versions.

# showUpdateAlert dialog with optional parameters and dialog dismissal
To change the appearance and behavior of the update dialog, you can `showUpdateAlert`
optionally provide `dialogTitle: String`, `dialogText: String`, `updateButtonText: String`, `allowDismissal: bool`,
`dismissButtonText: String`, and `dismissAction: VoidCallback` parameters.
```Dart
    appUpdate.showUpdateAlert(context: context,
                                dialogTitle: dialogTitle,
                                dialogText: dialogText,
                                updateButtonText: updateButtonText,
                                allowDismissal: true,
                                dismissButtonText: dismissButtonText,
                                dismissAction: dismissAction
                              );
```

# showUpdateAlert dialog without dialog dismissal
To change the appearance and behavior of the update dialog, you can `showUpdateAlert`
optionally provide `dialogTitle: String`, `dialogText: String`, `updateButtonText: String`, `allowDismissal: bool`, parameters.
```Dart
    appUpdate.showUpdateAlert(context: context,
                                dialogTitle: dialogTitle,
                                dialogText: dialogText,
                                updateButtonText: updateButtonText,
                                allowDismissal: false,
                              );
```

# showUpdateAlert dialog with optional VoidCallback
When dialog dismissal is allowed and to perform/ execute any functions on `dismissButtonText` or on dialog dismiss use the `dismissAction`.
```Dart
    void callBackFunction() {
      print('dismissAction called');
    }   

    appUpdate.showUpdateAlert(context: context,
                                dialogTitle: dialogTitle,
                                dialogText: dialogText,
                                updateButtonText: updateButtonText,
                                allowDismissal: true,
                                dismissButtonText: dismissButtonText,
                                dismissAction: callBackFunction()
                              );
```

# Force update dialog
Calling `forceAppVersion` with your app's greater version will check if the app can be updated, and will automatically display a platform-specific alert that the user can use to go to the app store.
```Dart
 final appUpdate = AppUpdate(
      iosPackageName: 'com.your.IOSpackage',
      androidPackageName: 'com.your.Androidpackage', 
      iOSAppStoreCountry: 'in',
      forceAppVersion: '1.0.1'
  );
```

# Platform Support 
```
ANDROID - ✅ Yes	                  
IOS     - ✅ Yes                   
LINUX   - ❌ No	                  
MACOS   - ❌ No	                 
WEB     - ❌ No	                  
WINDOWS - ❌ No	                  
```
# Support my work
<a href="https://www.buymeacoffee.com/bastevinod" target="_blank"><img src="https://cdn.buymeacoffee.com/buttons/default-orange.png" alt="Buy Me A Coffee" height="41" width="174"></a>

# License
```
Copyright [2022] [Vinod Baste]

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
```


CHANGELOG.md
## 1.5.5 - 03/11/22

* API includes optional parameters for `showUpdateAlert` method. Support for iOS and Android.


lib/app_updates.dart
library app_updates;

import 'dart:convert';
import 'dart:io';

import 'package:collection/collection.dart';
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:html/parser.dart';
import 'package:package_info_plus/package_info_plus.dart';
import 'package:url_launcher/url_launcher.dart';
import 'package:http/http.dart' as http;

///Info about the most recent and current versions of the software
///that are accessible through the Apple App Store or Google Play Store.
class AppVersionStatus {
  final String appVersion;
  final String storeVersion;
  final String storeLink;
  final String? releaseNotes;

  /// Returns `true` if the store version of the application is greater than the local version.
  bool get isUpdateAvailable {
    final appLocalVersion = appVersion.split('.').map(int.parse).toList();
    final appStoreVersion = storeVersion.split('.').map(int.parse).toList();

    //Only one comparison needs to return "true" in order to establish that the appStoreVersion version
    // is greater than the appLocalVersion version.
    // This is because each subsequent field in the version notation is less important than
    // the one before it.
    for (var i = 0; i < appStoreVersion.length; i++) {
      if (appStoreVersion[i] > appLocalVersion[i]) {
        return true;
      }
      if (appLocalVersion[i] > appStoreVersion[i]) {
        return false;
      }
    }
    return false;
  }

  AppVersionStatus._({
    required this.appVersion,
    required this.storeVersion,
    required this.storeLink,
    this.releaseNotes,
  });
}

/// An optional value that will force the plugin to always return [forceAppVersion]
/// as the value of [storeVersion]. This can be useful to test the plugin's behavior
/// before publishing a new version.
class AppUpdates {
  final String? iosPackageName;
  final String? androidPackageName;
  final String? iosAppStoreCountry;
  final String? forceAppVersion;

  AppUpdates({
    this.androidPackageName,
    this.iosPackageName,
    this.iosAppStoreCountry,
    this.forceAppVersion,
  });

  ///It then shows a platform-specific alert with buttons to ignore the update
  ///alert or access the app store after checking the version status.
  showUpdateAlert({required BuildContext context}) async {
    final AppVersionStatus? versionStatus = await _getVersionStatus();
    if (versionStatus != null && versionStatus.isUpdateAvailable) {
      _showUpdateDialog(context: context, versionStatus: versionStatus);
    }
  }

  /// This retrieves the data and verifies the version status.
  /// If you wish to use the data in a different way or display a custom alert, this is beneficial.
  Future<AppVersionStatus?> _getVersionStatus() async {
    PackageInfo packageInfo = await PackageInfo.fromPlatform();
    if (Platform.isIOS) {
      return _iosStoreVersion(packageInfo);
    } else if (Platform.isAndroid) {
      return _androidStoreVersion(packageInfo);
    } else {
      debugPrint(
          'The target platform "${Platform.operatingSystem}" is not yet supported by this package.');
    }
    return null;
  }

  /// This function attempts to clean local version strings so they match the MAJOR.MINOR.PATCH
  /// versioning pattern, so they can be properly compared with the store version.
  String _getCleanVersion(String version) =>
      RegExp(r'\d+\.\d+\.\d+').stringMatch(version) ?? '0.0.0';

  /// iOS info is fetched by using the iTunes lookup API, which returns a
  /// JSON document.
  Future<AppVersionStatus?> _iosStoreVersion(PackageInfo packageInfo) async {
    final id = iosPackageName ?? packageInfo.packageName;
    final parameters = {"bundleId": id};
    if (iosAppStoreCountry != null) {
      parameters.addAll({"country": iosAppStoreCountry!});
    }
    var uri = Uri.https("itunes.apple.com", "/lookup", parameters);
    final response = await http.get(uri);
    if (response.statusCode != 200) {
      debugPrint('Failed to query iOS App Store');
      return null;
    }
    final jsonObj = json.decode(response.body);
    final List results = jsonObj['results'];
    if (results.isEmpty) {
      debugPrint('Can\'t find an app in the App Store with the id: $id');
      return null;
    }
    return AppVersionStatus._(
      appVersion: _getCleanVersion(packageInfo.version),
      storeVersion:
      _getCleanVersion(forceAppVersion ?? jsonObj['results'][0]['version']),
      storeLink: jsonObj['results'][0]['trackViewUrl'],
      releaseNotes: jsonObj['results'][0]['releaseNotes'],
    );
  }

  /// Android info is fetched by parsing the html of the app store page.
  Future<AppVersionStatus?> _androidStoreVersion(
      PackageInfo packageInfo) async {
    final id = androidPackageName ?? packageInfo.packageName;
    final uri =
    Uri.https("play.google.com", "/store/apps/details", {"id": id, "hl": "en"});
    final response = await http.get(uri);
    if (response.statusCode != 200) {
      debugPrint('Can\'t find an app in the Play Store with the id: $id');
      return null;
    }
    final document = parse(response.body);

    String storeVersion = '0.0.0';
    String? releaseNotes;

    final additionalInfoElements = document.getElementsByClassName('hAyfc');
    if (additionalInfoElements.isNotEmpty) {
      final versionElement = additionalInfoElements.firstWhere(
            (elm) => elm.querySelector('.BgcNfc')!.text == 'Current Version',
      );
      storeVersion = versionElement.querySelector('.htlgb')!.text;

      final sectionElements = document.getElementsByClassName('W4P4ne');
      final releaseNotesElement = sectionElements.firstWhereOrNull(
            (elm) => elm.querySelector('.wSaTQd')!.text == 'What\'s New',
      );
      releaseNotes = releaseNotesElement
          ?.querySelector('.PHBdkd')
          ?.querySelector('.DWPxHb')
          ?.text;
    } else {
      final scriptElements = document.getElementsByTagName('script');
      final infoScriptElement = scriptElements.firstWhere(
            (elm) => elm.text.contains('key: \'ds:4\''),
      );

      final param = infoScriptElement.text.substring(20, infoScriptElement.text.length - 2)
          .replaceAll('key:', '"key":')
          .replaceAll('hash:', '"hash":')
          .replaceAll('data:', '"data":')
          .replaceAll('sideChannel:', '"sideChannel":')
          .replaceAll('\'', '"');
      final parsed = json.decode(param);
      final data =  parsed['data'];

      storeVersion = data[1][2][140][0][0][0];
      releaseNotes = data[1][2][144][1][1];
    }

    return AppVersionStatus._(
      appVersion: _getCleanVersion(packageInfo.version),
      storeVersion: _getCleanVersion(forceAppVersion ?? storeVersion),
      storeLink: uri.toString(),
      releaseNotes: releaseNotes,
    );
  }
  /// Shows the user a platform-specific alert about the app update. The user
  /// can dismiss the alert or proceed to the app store.
  ///
  /// To change the appearance and behavior of the update dialog, you can
  /// optionally provide [dialogTitle], [dialogText], [updateButtonText],
  /// [dismissButtonText], and [dismissAction] parameters.
  void _showUpdateDialog({
    required BuildContext context,
    required AppVersionStatus versionStatus,
    String dialogTitle = 'Update Available',
    String? dialogText,
    String updateButtonText = 'Update',
    bool allowDismissal = true,
    String dismissButtonText = 'Maybe Later',
    VoidCallback? dismissAction,
  }) async {
    final dialogTitleWidget = Text(dialogTitle);
    final dialogTextWidget = Text(
      dialogText ??
          'Update your app now to latest version and give it a spin!',
    );

    final updateButtonTextWidget = Text(updateButtonText);
    updateAction() {
      _launchAppStore(versionStatus.storeLink);
      if (allowDismissal) {
        Navigator.of(context, rootNavigator: true).pop();
      }
    }

    List<Widget> actions = [
      Platform.isAndroid
          ? TextButton(
        onPressed: updateAction,
        child: updateButtonTextWidget,
      )
          : CupertinoDialogAction(
        onPressed: updateAction,
        child: updateButtonTextWidget,
      ),
    ];

    if (allowDismissal) {
      final dismissButtonTextWidget = Text(dismissButtonText);
      dismissAction = dismissAction ??
              () => Navigator.of(context, rootNavigator: true).pop();
      actions.add(
        Platform.isAndroid
            ? TextButton(
          onPressed: dismissAction,
          child: dismissButtonTextWidget,
        )
            : CupertinoDialogAction(
          onPressed: dismissAction,
          child: dismissButtonTextWidget,
        ),
      );
    }

    await showDialog(
      context: context,
      barrierDismissible: allowDismissal,
      builder: (BuildContext context) {
        return WillPopScope(
            child: Platform.isAndroid
                ? AlertDialog(
              title: dialogTitleWidget,
              content: dialogTextWidget,
              actions: actions,
            )
                : CupertinoAlertDialog(
              title: dialogTitleWidget,
              content: dialogTextWidget,
              actions: actions,
            ),
            onWillPop: () => Future.value(allowDismissal));
      },
    );
  }

  /// Launches the Apple App Store or Google Play Store page for the app.
  Future<void> _launchAppStore(String appStoreLink) async {
    debugPrint(appStoreLink);
    if (await canLaunchUrl(Uri.parse(appStoreLink))) {
      await canLaunchUrl(Uri.parse(appStoreLink));
    } else {
      throw 'Could not launch appStoreLink';
    }
  }
}


example/lib/main.dart

import 'package:app_update_alert/appApdateAlert.dart';
import 'package:flutter/material.dart';


void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key});

  @override
  MyHomePageState createState() => MyHomePageState();
}

class MyHomePageState extends State<MyHomePage> {
  @override
  void initState() {
    super.initState();

    // Instantiate AppUpdate manager object (Using GCP Console app as example)
    final appUpdate = AppUpdates(
        iosPackageName: 'com.google.Vespa',
        androidPackageName: 'com.google.android.apps.cloudconsole',
        iosAppStoreCountry: 'in'
    );

    // Instantiate AppUpdate manager object with forceAppVersion
    final forceAppUpdate = AppUpdates(
        iosAppStoreCountry: 'in',
        forceAppVersion: '1.0.1'
    );

    // You can let the plugin handle fetching the status and showing a dialog,
    // or you can add version and forceUpdate the status to display the dialog.
    const forceUpdate = true;

    if (forceUpdate) {
      showUpdateDialog(forceAppUpdate);
    } else {
      showUpdateDialog(appUpdate);
    }
  }

  showUpdateDialog(AppUpdates appUpdate) {
    appUpdate.showUpdateAlert(context: context);
  }


  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text("Example App"),
      ),
    );
  }
}

lib/appApdateAlert.dart
library inapp_updates;

import 'dart:convert';
import 'dart:io';

import 'package:collection/collection.dart';
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:html/parser.dart';
import 'package:http/http.dart' as http;
import 'package:package_info_plus/package_info_plus.dart';
import 'package:url_launcher/url_launcher.dart';

///Info about the most recent and current versions of the software
///that are accessible through the Apple App Store or Google Play Store.
class AppVersionStatus {
  final String appVersion;
  final String storeVersion;
  final String storeLink;
  final String? releaseNotes;

  AppVersionStatus._({
    required this.appVersion,
    required this.storeVersion,
    required this.storeLink,
    this.releaseNotes,
  });

  /// Returns `true` if the store version of the application is greater than the local version.
  bool get isUpdateAvailable {
    final appLocalVersion = appVersion.split('.').map(int.parse).toList();
    final appStoreVersion = storeVersion.split('.').map(int.parse).toList();

    //Only one comparison needs to return "true" in order to establish that the appStoreVersion version
    // is greater than the appLocalVersion version.
    // This is because each subsequent field in the version notation is less important than
    // the one before it.
    for (var i = 0; i < appStoreVersion.length; i++) {
      if (appStoreVersion[i] > appLocalVersion[i]) {
        return true;
      }
      if (appLocalVersion[i] > appStoreVersion[i]) {
        return false;
      }
    }
    return false;
  }
}

/// An optional value that will force the plugin to always return [forceAppVersion]
/// as the value of [storeVersion]. This can be useful to test the plugin's behavior
/// before publishing a new version.
class AppUpdates {
  final String? iosPackageName;
  final String? androidPackageName;
  final String? iosAppStoreCountry;
  final String? forceAppVersion;

  AppUpdates({
    this.androidPackageName,
    this.iosPackageName,
    this.iosAppStoreCountry,
    this.forceAppVersion,
  });

  /// Shows the user a platform-specific alert about the app update. The user
  /// can dismiss the alert or proceed to the app store.
  ///
  /// To change the appearance and behavior of the update dialog, you can
  /// optionally provide [dialogTitle], [dialogText], [updateButtonText], [allowDismissal],
  /// [dismissButtonText], and [dismissAction] parameters.
  showUpdateAlert({required BuildContext context,
    String? dialogTitle,
    String? dialogText,
    String? updateButtonText,
    bool? allowDismissal,
    String? dismissButtonText,
    VoidCallback? dismissAction}) async {

    final AppVersionStatus? versionStatus = await _getVersionStatus();

    if (versionStatus != null && versionStatus.isUpdateAvailable) {

      _showUpdateDialog(
          context: context,
          versionStatus: versionStatus,
          dialogTitle: dialogTitle,
          dialogText: dialogText,
          updateButtonText: updateButtonText,
          allowDismissal: allowDismissal,
          dismissButtonText: dismissButtonText,
          dismissAction: dismissAction);
    }
  }

  /// This retrieves the data and verifies the version status.
  /// If you wish to use the data in a different way or display a custom alert, this is beneficial.
  Future<AppVersionStatus?> _getVersionStatus() async {
    PackageInfo packageInfo = await PackageInfo.fromPlatform();
    if (Platform.isIOS) {
      return _iosStoreVersion(packageInfo);
    } else if (Platform.isAndroid) {
      return _androidStoreVersion(packageInfo);
    } else {
      debugPrint(
          'The target platform "${Platform.operatingSystem}" is not yet supported by this package.');
    }
    return null;
  }

  /// This function attempts to clean local version strings so they match the MAJOR.MINOR.PATCH
  /// versioning pattern, so they can be properly compared with the store version.
  String _getCleanVersion(String version) =>
      RegExp(r'\d+\.\d+\.\d+').stringMatch(version) ?? '0.0.0';

  /// iOS info is fetched by using the iTunes lookup API, which returns a
  /// JSON document.
  Future<AppVersionStatus?> _iosStoreVersion(PackageInfo packageInfo) async {
    final id = iosPackageName ?? packageInfo.packageName;
    final parameters = {"bundleId": id};
    if (iosAppStoreCountry != null) {
      parameters.addAll({"country": iosAppStoreCountry!});
    }
    var uri = Uri.https("itunes.apple.com", "/lookup", parameters);
    final response = await http.get(uri);
    if (response.statusCode != 200) {
      debugPrint('Failed to query iOS App Store');
      return null;
    }
    final jsonObj = json.decode(response.body);
    final List results = jsonObj['results'];
    if (results.isEmpty) {
      debugPrint('Can\'t find an app in the App Store with the id: $id');
      return null;
    }
    return AppVersionStatus._(
      appVersion: _getCleanVersion(forceAppVersion ?? packageInfo.version),
      storeVersion:
          _getCleanVersion(forceAppVersion ?? jsonObj['results'][0]['version']),
      storeLink: jsonObj['results'][0]['trackViewUrl'],
      releaseNotes: jsonObj['results'][0]['releaseNotes'],
    );
  }

  /// Android info is fetched by parsing the html of the app store page.
  Future<AppVersionStatus?> _androidStoreVersion(
      PackageInfo packageInfo) async {
    final id = androidPackageName ?? packageInfo.packageName;
    final uri = Uri.https(
        "play.google.com", "/store/apps/details", {"id": id, "hl": "en"});
    final response = await http.get(uri);
    if (response.statusCode != 200) {
      debugPrint('Can\'t find an app in the Play Store with the id: $id');
      return null;
    }
    final document = parse(response.body);

    String storeVersion = '0.0.0';
    String? releaseNotes;

    final additionalInfoElements = document.getElementsByClassName('hAyfc');
    if (additionalInfoElements.isNotEmpty) {
      final versionElement = additionalInfoElements.firstWhere(
        (elm) => elm.querySelector('.BgcNfc')!.text == 'Current Version',
      );
      storeVersion = versionElement.querySelector('.htlgb')!.text;

      final sectionElements = document.getElementsByClassName('W4P4ne');
      final releaseNotesElement = sectionElements.firstWhereOrNull(
        (elm) => elm.querySelector('.wSaTQd')!.text == 'What\'s New',
      );
      releaseNotes = releaseNotesElement
          ?.querySelector('.PHBdkd')
          ?.querySelector('.DWPxHb')
          ?.text;
    } else {
      final scriptElements = document.getElementsByTagName('script');
      final infoScriptElement = scriptElements.firstWhere(
        (elm) => elm.text.contains('key: \'ds:4\''),
      );

      final param = infoScriptElement.text
          .substring(20, infoScriptElement.text.length - 2)
          .replaceAll('key:', '"key":')
          .replaceAll('hash:', '"hash":')
          .replaceAll('data:', '"data":')
          .replaceAll('sideChannel:', '"sideChannel":')
          .replaceAll('\'', '"');
      final parsed = json.decode(param);
      final data = parsed['data'];

      storeVersion = data[1][2][140][0][0][0];
      releaseNotes = data[1][2][144][1][1];
    }

    return AppVersionStatus._(
      appVersion: _getCleanVersion(forceAppVersion ?? packageInfo.version),
      storeVersion: _getCleanVersion(forceAppVersion ?? storeVersion),
      storeLink: uri.toString(),
      releaseNotes: releaseNotes,
    );
  }

  /// Shows the user a platform-specific alert about the app update. The user
  /// can dismiss the alert or proceed to the app store.
  ///
  /// To change the appearance and behavior of the update dialog, you can
  /// optionally provide [dialogTitle], [dialogText], [updateButtonText],
  /// [dismissButtonText], and [dismissAction] parameters.
  void _showUpdateDialog({
    required BuildContext context,
    required AppVersionStatus versionStatus,
    String? dialogTitle,
    String? dialogText,
    String? updateButtonText,
    bool? allowDismissal,
    String? dismissButtonText,
    VoidCallback? dismissAction,
  }) async {
    final dialogTitleWidget = Text(dialogTitle ?? 'Update Available');
    final dialogTextWidget = Text(
      dialogText ?? 'Update your app now to latest version and give it a spin!',
    );

    final updateButtonTextWidget = Text(updateButtonText ?? 'Update');
    updateAction() {
      _launchAppStore(versionStatus.storeLink);
      if (allowDismissal ?? true) {
        Navigator.of(context, rootNavigator: true).pop();
      }
    }

    List<Widget> actions = [
      Platform.isAndroid
          ? TextButton(
              onPressed: updateAction,
              child: updateButtonTextWidget,
            )
          : CupertinoDialogAction(
              onPressed: updateAction,
              child: updateButtonTextWidget,
            ),
    ];

    if (allowDismissal ?? true) {
      final dismissButtonTextWidget = Text(dismissButtonText ?? 'Maybe Later');
      dismissAction = dismissAction ??
          () => Navigator.of(context, rootNavigator: true).pop();
      actions.add(
        Platform.isAndroid
            ? TextButton(
                onPressed: dismissAction,
                child: dismissButtonTextWidget,
              )
            : CupertinoDialogAction(
                onPressed: dismissAction,
                child: dismissButtonTextWidget,
              ),
      );
    }

    await showDialog(
      context: context,
      barrierDismissible: allowDismissal ?? true,
      builder: (BuildContext context) {
        return WillPopScope(
            child: Platform.isAndroid
                ? AlertDialog(
                    title: dialogTitleWidget,
                    content: dialogTextWidget,
                    actions: actions,
                  )
                : CupertinoAlertDialog(
                    title: dialogTitleWidget,
                    content: dialogTextWidget,
                    actions: actions,
                  ),
            onWillPop: () => Future.value(allowDismissal));
      },
    );
  }

  /// Launches the Apple App Store or Google Play Store page for the app.
  Future<void> _launchAppStore(String appStoreLink) async {
    debugPrint(appStoreLink);
    if (await canLaunchUrl(Uri.parse(appStoreLink))) {
      await canLaunchUrl(Uri.parse(appStoreLink));
    } else {
      throw 'Could not launch appStoreLink';
    }
  }
}
